harvester/harvester
# ResourceQuota add insufficient resource annotation and fixed vm when in starting

**Problem:**
- From a UX perspective, it appears that a user attempts to power on the VM, and nothing happens. 
- When any VM starts simultaneously (batch create VMs) and exceeds ResourceQuota, properly have a few VMs blocked in Starting status.
- If the namespace ResourceQuota update's value is less than the used quota, the VMs can't migrate.

**Solution:**
- Enhancement UX that add annotation of insufficient resource(`harvesterhci.io/insufficientResourceQuota: xxxxx`), and UI shows messages, once VM is running, will be cleaned up the annotation.
- If the VM is in Starting status, that VM re-queue after five seconds until non in Starting status.
- Improve the namespace ResourceQuota validator so that if the update's value is less than the used quota, then reject the update request.

**Related Issue:**
https://github.com/harvester/harvester/issues/3944
https://github.com/harvester/harvester/pull/3829#issuecomment-1556382198
https://github.com/harvester/harvester/issues/3974

**Test plan:**
Insufficient Resource Quota Annotation:
1. Batch create VMs(or batch to start), VMs resource must exceed the ResourceQuota.
2. Check the start failed VMs' annotation `harvesterhci.io/insufficientResourceQuota: xxxxx`.
3. After stopping the running VMs, start the VMs of `harvesterhci.io/insufficientResourceQuota: xxxxx`.
4. Check the running VMs to clean up the annotation.

Check block the Starting status:
1. Batch creates VMs(or batch to start), and VMs resource must exceed the ResourceQuota.
2. Check if any VMs are still in Starting status.

Namespace ResourceQuota validator:
https://github.com/harvester/harvester/issues/3974#issue-1725712196
